sources/api/thar-be-updates/src/main.rs

```diff
@@ -0,0 +1,427 @@
+                        }
+                    }
+                    write_update_status(&update_status)?;
+                    unlock(&lockfile)?;
+                }
+                Err(e) => {
+                    unlock(&lockfile)?;
+                    eprintln!("{}", e);
+                    return error::Fork.fail();
+                }
+            }
+        }
+        UpdateCommand::PrepareUpdate => {
+            lock_exclusive(&lockfile)?;
+            let mut update_status = get_update_status()?;
+            match update_status.update_state() {
+                UpdateState::Available | UpdateState::Staged => {
+                    // Make sure the chosen update exists
+                    let _ = update_status.chosen_update().context(error::ChosenUpdate)?;
+                    match fork() {
+                        Ok(ForkResult::Parent { child, .. }) => {
+                            debug!("forked child pid: {}", child);
+                            // Return immediately as the parent.
+                            // Parent's lockfile fd will close but child will still have a duplicate fd
+                            return Ok(());
+                        }
+                        Ok(ForkResult::Child) => {
+                            let result = prepare_update(&mut update_status);
+                            // Transition the update state appropriately
+                            if let UpdateState::Available = update_status.update_state() {
+                                if result.is_ok() {
+                                    update_status.to_staged()?;
+                                }
+                            }
+                            write_update_status(&update_status)?;
+                            unlock(&lockfile)?;
+                        }
+                        Err(e) => {
+                            unlock(&lockfile)?;
+                            eprintln!("{}", e);
+                            return error::Fork.fail();
+                        }
+                    }
+                }
+                UpdateState::Idle | UpdateState::Ready => {
+                    unlock(&lockfile)?;
+                    return error::DisallowCommand {
+                        command: UpdateCommand::PrepareUpdate,
+                        state: update_status.update_state().to_owned(),
+                    }
+                    .fail();
+                }
+            }
+        }
+        // Only allowed when state is 'Staged'
+        UpdateCommand::ActivateUpdate => {
+            lock_exclusive(&lockfile)?;
+            let mut update_status = get_update_status()?;
+            match update_status.update_state() {
+                UpdateState::Staged => {
+                    // Make sure there's an update image written to the inactive partition
+                    let _ = update_status
+                        .staging_partition()
+                        .context(error::StagingPartition)?;
```

Comment:
```suggestion
                    ensure!(update_status.staging_partition_ready(),
                            error::StagingPartition)?;
```

Reply:
Or alternatively:

```
ensure!(update_status.can_activate_update, error::StagingPartition)?;
activate_update(&mut update_status)?;
```

Where `update_status.to_ready()` happens inside `activate_update()`. The `ensure!` could go there too.